### ISSUE TITLE
Native driver crashed

### ISSUE THREAD

**Issue by ncordon** [bblfsh/cpp-driver#54]
From [bblfsh logs](https://github.com/bblfsh/bblfshd/issues/298) we got the error `native driver crashed`

<details>
<summary> Files that produced the error </summary>

```
udoc/headerfile.h
InputModuleOIS/Exports.cpp
release/src-rt/linux/linux-2.6/scripts/squashfs/lzma/C/7zip/Compress/Branch/x86.h
LibraryModule/StableHeaders.cpp
src/others/squashfs-4.0-realtek/lzma/CPP/7zip/Bundles/Alone7z/StdAfx.cpp
core/md5.h
ProtocolUtilities/LoginCredentials.h
PhysicsModule/CollisionShapeUtils.h
Foundation/ComponentManager.cpp
rosapps/applications/sysutils/utils/sdkparse/skip_ws.cpp
Engine/Source/Runtime/Android/AndroidRuntimeSettings/Private/AndroidRuntimeSettingsModule.cpp
gnu/dist/binutils/ld/testsuite/ld-cdtest/cdtest-bar.cc
encodings/gbk.h
TestModule/TestService.cpp
Templates/TP_Vehicle/Source/TP_Vehicle/TP_Vehicle.cpp
PythonScriptModule/TransformDecorator.h
SceneManager/StableHeaders.cpp
TelepathyIMModule/CommunicationUI/ConfigHelper.cpp
UiModule/Inworld/Notifications/NotificationLogWidget.h
ld/testsuite/ld-cdtest/cdtest-bar.cc
src/Core/TundraProtocolModule/StableHeaders.cpp
TelepathyIMModule/CommunicationUI/FriendListItem.cpp
sketches/Trident/CModule.cpp
smtp/smtpauth.h
Foundation/QtInputMouseEvent.cpp
encodings/cp.h
UiModule/Common/UiVisibilityAction.cpp
PythonScriptModule/TransformDecorator.cpp
EnvironmentModule/StableHeaders.cpp
smtp/smtpparser.h
Engine/Source/Runtime/Core/Private/HAL/IPlatformFileLogWrapper.cpp
src/EntityComponents/EC_Ruler/StableHeaders.cpp
RexLogicModule/NotificationWidget.h
WorldBuildingModule/StableHeaders.cpp
db/pgmessage.h
Core/CoreStableHeaders.cpp
CAVEStereoModule/CAVEManager.cpp
Tests/Tutorial/Step3/tutorial.cxx
test/SemaCXX/warn-sysheader-macro.cpp
UiModule/Inworld/ControlPanel/BackdropWidget.h
src/Core/ECEditorModule/LineEditPropertyFactory.cpp
PythonScriptModule/QEC_Prim.cpp
imap/handlers/rename.h
Engine/Source/Developer/AITestSuite/Private/TestLogger.cpp
RexLogicModule/EventHandlers/SceneEventHandler.h
Engine/Source/Runtime/InputCore/Private/InputCoreModule.cpp
ReXLogic/EC_OpenSimAvatar.cpp
Engine/Source/Runtime/Sockets/Private/MultichannelTCP.cpp
CommunicationModule/CommunicationUI/FriendListItem.cpp
kpresenter/rotationdialogimpl.cc
```

</details>

**Comment by kuba--** [comment]
I don't know if it's relevant, but you can crash the driver with following file:
https://github.com/microsoft/clang-1/blob/master/test/SemaCXX/cxx0x-initializer-stdinitializerlist.cpp

You'll get following error:
```
      --- FAIL: TestCppDriver/native/cxx0x-initializer-stdinitializerlist (1.94s)
            require.go:794: 
                	Error Trace:	fixtures.go:198
                	Error:      	Received unexpected error:
                	            	driver failure: unexpected end of JSON input
                	            	
                	            	github.com/bblfsh/sdk/v3/driver/native.(*Driver).Parse
                	            		/go/src/github.com/bblfsh/cpp-driver/vendor/github.com/bblfsh/sdk/v3/driver/native/native.go:312
                	            	github.com/bblfsh/sdk/v3/driver/fixtures.(*Suite).testFixturesNative.func1
                	            		/go/src/github.com/bblfsh/cpp-driver/vendor/github.com/bblfsh/sdk/v3/driver/fixtures/fixtures.go:189
                	            	testing.tRunner
                	            		/usr/local/go/src/testing/testing.go:865
                	            	runtime.goexit
                	            		/usr/local/go/src/runtime/asm_amd64.s:1337
                	Test:       	TestCppDriver/native/cxx0x-initializer-stdinitializerlist
    --- FAIL: TestCppDriver/uast (0.01s)
        --- FAIL: TestCppDriver/uast/cxx0x-initializer-stdinitializerlist (0.01s)
            require.go:794: 
                	Error Trace:	fixtures.go:86
                	            				fixtures.go:255
                	Error:      	Received unexpected error:
                	            	open ../../fixtures/cxx0x-initializer-stdinitializerlist.cpp.native: no such file or directory
                	Test:       	TestCppDriver/uast/cxx0x-initializer-stdinitializerlist
    --- FAIL: TestCppDriver/semantic (0.00s)
        --- FAIL: TestCppDriver/semantic/cxx0x-initializer-stdinitializerlist (0.00s)
            require.go:794: 
                	Error Trace:	fixtures.go:86
                	            				fixtures.go:255
                	Error:      	Received unexpected error:
                	            	open ../../fixtures/cxx0x-initializer-stdinitializerlist.cpp.native: no such file or directory
                	Test:       	TestCppDriver/semantic/cxx0x-initializer-stdinitializerlist
FAIL
```

**Comment by kuba--** [comment]
For mentioned file(`cxx0x-initializer-stdinitializerlist.cpp`) CDT parser fails with errors:
```
 "tech.sourced.babelfish.DriverResponse.ResponseSendException",
"Error serializing the AST to JSON: 
com.fasterxml.jackson.databind.JsonMappingException: org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to 
org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer (through reference chain: tech.sourced.babelfish.DriverResponse[\"ast\"])",

"tech.sourced.babelfish.DriverResponse$ResponseSendException: 
com.fasterxml.jackson.databind.JsonMappingException: 
org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to 
org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer (through reference chain: tech.sourced.babelfish.DriverResponse[\"ast\"])\n\tat 

tech.sourced.babelfish.DriverResponse.send(DriverResponse.java:53)\n\tat 
tech.sourced.babelfish.Main.process(Main.java:53)\n\tat 
tech.sourced.babelfish.Main.main(Main.java:13)\nCaused by: 
com.fasterxml.jackson.databind.JsonMappingException: 
org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to 
org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer (through reference chain: tech.sourced.babelfish.DriverResponse[\"ast\"])\n\tat 
com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:388)\n\tat 
com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:348)\n\tat 
com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:343)\n\tat 
com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:697)\n\tat 
com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:155)\n\tat 
com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:292)\n\tat 
com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:2509)\n\tat 
tech.sourced.babelfish.TranslationUnitJSONMapper.writeValue(TranslationUnitJSONMapper.java:41)\n\tat 
tech.sourced.babelfish.DriverResponse.send(DriverResponse.java:48)\n\t... 2 more\nCaused by: 
java.lang.ClassCastException: org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to 
org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createAutoType(CPPVisitor.java:2269)\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createType(CPPVisitor.java:2147)\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createType(CPPVisitor.java:2115)\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createType(CPPVisitor.java:1915)\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTNewExpression.getEvaluation(CPPASTNewExpression.java:289)\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPEvaluation.getType(CPPEvaluation.java:61)\n\tat 
org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTNewExpression.getExpressionType(CPPASTNewExpression.java:308)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:115)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serialize(TranslationUnitSerializer.java:70)\n\tat 
tech.sourced.babelfish.TranslationUnitSerializer.serialize(TranslationUnitSerializer.java:44)\n\tat 
com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:704)\n\tat 
com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:689)\n\t... 7 more\n"
```

what can be the reason of driver's crash

**Comment by kuba--** [comment]
Found a snippet which caused an error of native parser (CDT):
```c++
void dangle() {
  new auto{1, 2, 3}; // expected-error {{cannot use list-initialization}}
  new std::initializer_list<int>{1, 2, 3}; // expected-warning {{at the end of the full-expression}}
}
```

```
"tech.sourced.babelfish.DriverResponse.ResponseSendException",
"Error serializing the AST to JSON: com.fasterxml.jackson.databind.JsonMappingException: org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer (through reference chain: tech.sourced.babelfish.DriverResponse[\"ast\"])",
        "tech.sourced.babelfish.DriverResponse$ResponseSendException: com.fasterxml.jackson.databind.JsonMappingException: org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer (through reference chain: tech.sourced.babelfish.DriverResponse[\"ast\"])\n\tat tech.sourced.babelfish.DriverResponse.send(DriverResponse.java:53)\n\tat tech.sourced.babelfish.Main.process(Main.java:53)\n\tat tech.sourced.babelfish.Main.main(Main.java:13)\nCaused by: com.fasterxml.jackson.databind.JsonMappingException: org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer (through reference chain: tech.sourced.babelfish.DriverResponse[\"ast\"])\n\tat com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:388)\n\tat com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:348)\n\tat com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:343)\n\tat com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:697)\n\tat com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:155)\n\tat com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:292)\n\tat com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:2509)\n\tat tech.sourced.babelfish.TranslationUnitJSONMapper.writeValue(TranslationUnitJSONMapper.java:41)\n\tat tech.sourced.babelfish.DriverResponse.send(DriverResponse.java:48)\n\t... 2 more\nCaused by: java.lang.ClassCastException: org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createAutoType(CPPVisitor.java:2269)\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createType(CPPVisitor.java:2147)\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createType(CPPVisitor.java:2115)\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPVisitor.createType(CPPVisitor.java:1915)\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTNewExpression.getEvaluation(CPPASTNewExpression.java:289)\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.semantics.CPPEvaluation.getType(CPPEvaluation.java:61)\n\tat org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTNewExpression.getExpressionType(CPPASTNewExpression.java:308)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:115)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serializeNode(TranslationUnitSerializer.java:141)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serialize(TranslationUnitSerializer.java:70)\n\tat tech.sourced.babelfish.TranslationUnitSerializer.serialize(TranslationUnitSerializer.java:44)\n\tat com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:704)\n\tat com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:689)\n\t... 7 more\n"
```

**Comment by kuba--** [comment]
I went through _some deep dive debugging_. I split the process into independent steps:
- serialize source file into `parse request`
- dissemble and debug CDT parser (cdtcore 6.5.0)
- test sdk decoder based on `parse response`
- debug sdk's `native.Parse` used by driver.

To make a long story short, I think the driver *does not* crash. The error message in logs is somehow misleading.

### TL;DR
The file which I mentioned (`cxx0x-initializer-stdinitializerlist.cpp`) is one of the files which fail the tests. I've split the file and found the section/function which is responsible for the failure:
```c++
void dangle() {
  new auto{1, 2, 3}; // expected-error {{cannot use list-initialization}}
  new std::initializer_list<int>{1, 2, 3}; // expected-warning {{at the end of the full-expression}}
}
```
As you can see an error is expected.
That's why our _CDT_ parser throws an exception: 
```
        "tech.sourced.babelfish.DriverResponse.ResponseSendException",
        "Error serializing the AST to JSON:
...
```
All the _casting_ errors come from the fact that we cannot _cast_ invalid expression `new auto{1, 2, 3};` to any _built-in_ expression.
At the end, CTD returns just an error payload with empty _AST_.:
```json
{
    "driver": "1.0.0",
    "language": "C++",
    "languageVersion": "14",
    "status": "fatal",
    "errors": [
...
    ],
  "ast": null
}
```
This payload is taken as an input to `bblfsh/sdk`. Because the `status` is `fatal` it wraps the CDT stack trace by `driver failure` error and returns the error what fails the test:
https://github.com/bblfsh/sdk/blob/master/driver/native/native.go#L309
https://github.com/bblfsh/sdk/blob/master/driver/native/native.go#L326

To recap, I don't think the driver is failing. In my opinion the errors are misleading. For cases where error is expected or when input is broken, we should fail, but the error messages, like:
`native driver crashed` or `driver failure` may look scary if see them in logs.

With any of files/snippets I could not reproduce any failure (technically speaking), so I suggest to close the issue, so far.

@dennwc, @ncordon

**Comment by dennwc** [comment]
@kuba-- Thanks for digging into it.

> In my opinion the errors are misleading.

They are, however, the assertion in the parser fails and it returns `fatal` which is correctly interpreted by SDK as "parser died".

> As you can see an error is expected.

In this case it's a syntax error and should be returned with a status `error` instead of `fatal`.

> we cannot cast invalid expression ... to any built-in expression.

This is actually surprising. I mean, the parsing definitely succeeded and a (partial?) AST is returned. But we fail to serialize it because of an invalid cast to  `ICPPASTConstructorInitializer` while the node type is `CPPASTInitializerList`. But what function makes this assertion? I think the most useful solution would be to fix the cast. In most cases the parser will have a `CPPASTProblem` node in the tree if it encounters any parsing errors.

> With any of files/snippets I could not reproduce any failure (technically speaking), so I suggest to close the issue, so far.

Also note that the original issues mentions other cases where driver "crashed" they might or might not be the same. So it's definitely too early to close this.

**Comment by kuba--** [comment]
@dennwc - let me describe what's going on inside CDT with more details (for the snippet):
```c++
void f () {
  new auto{1, 2, 3}; // expected-error {{cannot use list-initialization}}
 // ...
}
```

First of all CDT throws an exception which is caught in `Main`. In `Main` every exception is wrapped and the function `trySendError` is called:
```java
// ...
} catch (IOException e) {
    return trySendError("A problem occurred while processing the petition: ", e);
}
```
The function calls `void sendError(Exception e, String errorString)` which overwrites the status to `fatal`:
```java
//...
 errors.add(sw.toString());
 status = "fatal";
// ...
```
It happens in Main, so we can introduce more exceptions, what let us distinguish fatals from errors (in theory).

Let's back to the exception and why that happens.
`CPPVisitor` recognizes `auto` expression and tries create an `auto` type for a `new` expression.
We get to the following statement:
```java
IASTInitializerClause[] arguments = ((ICPPASTConstructorInitializer)initializer).getArguments();
```
So, as you can see it should be a _constructor_ (that's why this explicit casting to `ICPPASTConstructorInitializer`), but `initializer` here does not implement `ICPPASTConstructorInitializer`, because it's `CPPASTInitializerList` and there is not `get/setArgument` methods. Btw:
`public class CPPASTInitializerList extends ASTNode implements ICPPASTInitializerList, IASTAmbiguityParent `.

In other words, this casting throws an exception.
Upgrading CDT to the newest version will not solve the problem, because this code didn't change much over time:
https://github.com/eclipse-cdt/cdt/blob/master/core/org.eclipse.cdt.core/parser/org/eclipse/cdt/internal/core/dom/parser/cpp/semantics/CPPVisitor.java#L2516

An option is also to modify the own fork or send PR to CDT (and wait for a merge), where we can try sth. like `instanceof` and return  ProblemType.CANNOT_DEDUCE_AUTO_TYPE`.

> Also note that the original issues mentions other cases where driver "crashed" they might > or might not be the same. So it's definitely too early to close this.

Crash error (`ErrDriverCrashed`) happens only when you get an empty response (error is `io.EOF`) and after restart the driver (`sdk/native.go`):
```go
	if err == io.EOF {
		if err := d.restart(); err != nil {
			return nil, err
		}
		// fail anyway - this request may have caused the crash
		err = ErrDriverCrashed.New()
	}
```
In most cases there is _driver failure_ (`ErrDriverFailure`)

**Comment by dennwc** [comment]
@kuba-- Thanks again for a detailed response!

> It happens in Main, so we can introduce more exceptions, what let us distinguish fatals from errors (in theory).

Yes, this is exactly what I had in mind. But can we catch this specific exception though and distinguish it from JSON encoding errors and wrong casts in our own code, for example?

> An option is also to modify the own fork or send PR to CDT (and wait for a merge), where we can try sth. like instanceof and return `ProblemType.CANNOT_DEDUCE_AUTO_TYPE`.

Can we catch this exception without a fork? Or does this function being called inside the JSON encoder (so we can't localize the exception)? If so, sending the patch upstream sounds like a good idea.

But I think we can start from `ProblemType.UNEXPECTED_NODE_TYPE` or something similar. For now we don't care specifically about the type of the node in question because we can localize the node by position anyway.

> Crash error (ErrDriverCrashed) happens only when you get an empty response

Yeah, I think this is exactly what happens in [other cases](https://github.com/bblfsh/cpp-driver/issues/54#issue-472247394) mentioned in this issue.

**Comment by kuba--** [comment]
> Yes, this is exactly what I had in mind. But can we catch this specific exception though and distinguish it from JSON encoding errors and wrong casts in our own code, for example?

Partly, because what you can do in `DriverResponse` is to catch:
```java
catch(com.fasterxml.jackson.databind.JsonMappingException e) {
            if (e.getCause() instanceof java.lang.ClassCastException) {
                // ...
            }
```
But parsing the message `org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTInitializerList cannot be cast to 
org.eclipse.cdt.core.dom.ast.cpp.ICPPASTConstructorInitializer` does not make sense (IMO).
I'm thinking to extend `ResponseSendException` adding status, so in this particular case we can pass status `error` (otherwise fatal).


Or even simpler, try to unwrap everything in `sendError` and if it's casting issue in `JsonMappingException` then directly set status to error (instead of fatal).
WDYT?

```java
        if (e instanceof ResponseSendException &&
            e.getCause() instanceof JsonMappingException &&
            e.getCause().getCause() instanceof java.lang.ClassCastException) {
            status = Status.error;
        } else {
            status = Status.fatal;
        }
```
---

Regarding a patch, replacing c-like cast by `instanceof` makes sense, but on the other hand, from CDT point of view, an exception is the right thing to do (for this case).

**Comment by dennwc** [comment]
@kuba-- The simple you mentioned looks good, but I'm concerned about catching _any_ failed casts, including our own. If possible, I'd like to distinguish our own cast failures (fatal) from CDT cast failures (error).

Also it would be awesome to continue JSON encoding even in presence of those cast exceptions, but that may deserve a separate issue.

**Comment by kuba--** [comment]
what do you mean our own? the cpp-driver's, jackson's?
anyway, for casting exceptions we can traverse stack and check if the root cause comes from cdt package

**Comment by ncordon** [comment]
Wow, thank you for debugging this @kuba-- . I prefer the unwrapping option.

Respect to closing the issue, I am in favour if it does not fail for any of the files in the list, since then we are not going to be able to reproduce anything
